File: 3142-longest-unequal-adjacent-groups-subsequence-ii/longest-unequal-adjacent-groups-subsequence-ii.go
func getWordsInLongestSubsequence(words []string, groups []int) []string {
	n := len(groups);
	dp := make([]int, n);
	parent := make([]int, n);

	for i := 0; i < n; i++ {
		dp[i] = 1;
		parent[i] = -1;
	}

	maxLen := 0;
	lastIndex := 0;

	for i := 0; i < n; i++ {
		for j := 0; j < i; j++ {
			if groups[i] != groups[j] && isHammingDistance(words[i], words[j]) && dp[i] < dp[j]+1 {
				dp[i] = dp[j] + 1;
				parent[i] = j;
			}
		}

		if dp[i] > maxLen {
			maxLen = dp[i];
			lastIndex = i;
		}
	}

	// Truy vết ngược từ lastIndex
	var result []string;
	for i := lastIndex; i != -1; i = parent[i] {
		result = append(result, words[i]);
	}

	// Đảo ngược kết quả
	for i, j := 0, len(result)-1; i < j; i, j = i+1, j-1 {
		result[i], result[j] = result[j], result[i];
	}

	return result;
}

func isHammingDistance(a, b string) bool {
	if len(a) != len(b) {
		return false;
	}

	diff := 0;
	for i := 0; i < len(a); i++ {
		if a[i] != b[i] {
			diff++;
			if diff > 1 {
				return false;
			}
		}
	}
    
	return diff == 1;
}

File: 3142-longest-unequal-adjacent-groups-subsequence-ii/longest-unequal-adjacent-groups-subsequence-ii.java
class Solution {
    public List<String> getWordsInLongestSubsequence(String[] words, int[] groups) {
        int n = groups.length;
        int[] dp = new int[n];
        int[] parent = new int[n];
        Arrays.fill(dp, 1);
        Arrays.fill(parent, -1);
        int maxi = 0;

        for (int i = 0; i < n; i++) {
            for (int j = 0; j < i; j++) {
                if (groups[i] != groups[j] &&
                        isHammingDistance(words[i], words[j]) &&
                        dp[i] < dp[j] + 1) {
                    dp[i] = dp[j] + 1;
                    parent[i] = j;
                }
            }

            if (dp[i] > maxi)
                maxi = dp[i];
        }

        List<String> result = new ArrayList<>();
        for (int i = 0; i < n; i++) {
            if (dp[i] == maxi) {
                while (i != -1) {
                    result.add(words[i]);
                    i = parent[i];
                }
                break;
            }
        }
        
        Collections.reverse(result);
        return result;
    }

    public boolean isHammingDistance(String a, String b) {
        if (a.length() != b.length())
            return false;

        int difCount = 0;
        for (int i = 0; i < a.length(); i++) {
            if (a.charAt(i) != b.charAt(i)) {
                difCount++;
                if (difCount > 1)
                    return false;
            }
        }

        return difCount == 1;
    }
}